sympy/matrices/random.py

```diff
@@ -0,0 +1,2343 @@
+
+    rank = dim - 1 if rank is None else rank
+    index = numbers_with_sum(dim - rank, dim)
+    spec = tuple((0, i) for i in index)
+    normal_form = jordan_normal(dim, spec=spec)
+    s = invertible(dim, scalars, units, length)
+    return s.inv() * normal_form * s
+
+
+def diagonalizable(dim,
+                   spec=_elementary_scalars,
+                   scalars=_elementary_scalars,
+                   units=_elementary_units,
+                   length=None):
+    r"""a square matrix n x n of a given rank
+
+    Explanation
+    ===========
+    Creates a square matrix of a given rank
+    which is diagonalizable.
+
+    The matrix will be a product
+
+    .. math::
+
+        \mathbf{S} \cdot \mathbf{D} \cdot \mathbf{S}^{-1}
+
+    of an :func:`invertible` matrix $\mathbf{S}$ and
+    a :func:`diagonal_normal` matrix of given rank
+    and eigenvalues (diagonal entries) from **spec**.
+
+    To obtain a matrix with *complex* *eigenvalues*
+    but non-complex entries,
+    e.g. $\lambda_{1/2}=a \pm b \ i$,
+    you better may use a product of an invertible matrix $\mathbf{S}$
+    with non-complex entries (**scalars**)
+    and where $\mathbf{D}$ is a non-complex block diagonal matrix
+    with a block $\mathbf{D}_z$
+
+    .. math::
+
+        \mathbf{D}_z = \left[\begin{array}{cc}a & b \\ -b & a\end{array}\right]
+
+    Examples
+    ========
+    >>> from sympy.core.random import seed
+    >>> from sympy.matrices.random import diagonalizable
+    >>> seed(1)
+
+    >>> diagonalizable(3)
+    Matrix([
+    [-1,  0, 0],
+    [-2, -3, 4],
+    [-2, -2, 3]])
+
+    >>> m = diagonalizable(3, spec=(1,2,2), scalars=(1,2,3), length=10)
+    >>> m
+    Matrix([
+    [ 4,  2, -6],
+    [-6, -4, 18],
+    [-1, -1,  5]])
+    >>> m.eigenvals(multiple=True)
+    [1, 2, 2]
+    >>> m.jordan_form(calc_transform=False)
+    Matrix([
+    [1, 0, 0],
+    [0, 2, 0],
+    [0, 0, 2]])
+
+    Parameters
+    ==========
+    dim : integer
+        dimension of matrix
+    spec : tuple or list of symbols (optional)
+        set of values of which scalars (diagonal entries) are choosen
+        see :func:`diagonal_normal`
+    scalars : tuple or list of symbols (optional)
+        see :func:`invertible`
+    units : tuple or list of symbols (optional)
+        see :func:`invertible`
+    length : integer (optional with default 2 * **dim**)
+        see :func:`invertible`
+
+    """
+
+    normal_form = diagonal_normal(dim, spec)
+    s = invertible(dim, scalars, units, length)
+    return s.inv() * normal_form * s
+
+
+def trigonalizable(dim,
+                   spec=_elementary_scalars,
+                   scalars=_elementary_scalars,
+                   units=_elementary_units,
+                   length=None):
+    r"""Creates a square matrix n x n of a given rank
+
+    Explanation
+    ===========
+    Creates a square matrix n x n of a given rank
+    which is trigonalizable (has echelon form)
+    and has eigenvalues from a given set.
```

Comment:
What exactly does trigonalizable mean? Every matrix has an echelon form.

How is this different from `square`?

Reply:
Yes, `trigonalizable` means it has an echelon form but with a fixed **spectrum** i.e. diagonal entries in echelon form. As given by `spec`. `square` gives a general square matrix which does only have general diagonal entries in echelon form. 

This function is very useful for generating matrix exercises in undergrad math courses.

Reply:
Do you have any source for this meaning of trigonalizable? When I search this word I get e.g.:
https://encyclopediaofmath.org/wiki/Trigonalizable_element
Translating that into our language it says that a square matrix over a field K is trigonalizable if all its eigenvalues lie in K rather than some algebraic extension of K. If we allow K to be algebraically closed (e.g. the complex numbers) then every square matrix is trigonalizable so there needs to be some notion of the field that we are working over for trigonalizable to be a meaningful property of a square matrix.

Here there is no mention of a field but I guess that these matrices are implicitly defined to be in a ring something like `Z[s1,s2,...][e1,e2,...]` where the `si` are the scalars and the `ei` are the spec (eigenvalues).

I think perhaps this docstring could be a bit clearer in defining what a trigonalizable matrix is. I'm not sure what the best language is but I found this confusing to read also because the meaning of trigonalizable is not obvious from a quick web search.

Reply:
_trigonalizable_ means for square matrix $A$ that it is similar to an upper triangular matrix $T$ see [Wikipedia](https://en.wikipedia.org/wiki/Triangular_matrix#Triangularisability), i.e. there is an $S \in GL(n)$ s.th. $S^{-1} T S = A$.

Same as a diagonizable matrix is similar to a diagonal matrix.

Here the field $k$ is actually given by the _spec_, _scalars_ and _units_ arguments. If all of them are elements of $k$.

Reply:
I have heard of the word `triangularizable` (the word used in wikipedia) before but `trigonalizable` is new to me. Is there a reason to prefer `trigonalizable`?

Reply:
Sorry, didn't realize the difference in spelling. _triangularizable_ is fine for me also I think _trigonalizable_ wouldn't be wrong. Nevertheless, I'll change it tomorrow.